llvm/lib/IR/RepoTicket.cpp

```diff
@@ -116,78 +125,135 @@ static GODigestState accumulateGOFinalDigestOrInitialDigestAndGetGODigestState(
+    ++VisitedCachedGOTimes;
+    LLVM_DEBUG(dbgs() << "Returning pre-computed hash for \"" << GO->getName()
+                      << "\"\n");
+    return std::make_tuple(true, NumVisited, TablePos->second);
+  }
+
+  MD5 GOHash = MD5();
 
   bool Inserted;
   typename GOStateMap::const_iterator StateIt;
-  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, Visited.size());
+  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, NumVisited);
   if (!Inserted) {
-    // If GO is visited, use the letter 'R' as the marker and use its state as
-    // the value.
-    GOHash.update('R');
-    GOHash.update(StateIt->second);
-    return true;
+    assert(NumVisited > StateIt->second);
+    // If GO is visited, mark as a Backref and use its state as the value.
+    GOHash.update(static_cast<char>(Tags::Backref));
+    GOHash.update(NumVisited - StateIt->second - 1U);
+    LLVM_DEBUG(dbgs() << "Hashing back reference to #" << StateIt->second
+                      << "\n");
+    MD5::MD5Result Digest;
+    GOHash.final(Digest);
+    return std::make_tuple(true, StateIt->second, Digest);
   }
 
-  bool Changed = false;
-  GOHash.update('T');
+  GOHash.update(static_cast<char>(Tags::GO));
   auto State = AccumulateGODigest(GO, GOHash, GOIMap);
-  Changed = Changed || State.Changed;
-  // Update the GO's hash using its contributions and dependencies.
-  auto &GOContributions = State.Contributions;
-  auto &GODependencies = State.Dependencies;
-  for (const GlobalObject *const G : llvm::concat<const GlobalObject *const>(
-           make_range(GOContributions.begin(), GOContributions.end()),
-           make_range(GODependencies.begin(), GODependencies.end()))) {
+  // Add the GO's contributions to its hash. Contributions can't form loops
+  // since 1) only global variables have contributions and 2) all contributions
+  // are functions.
+  bool Changed = State.Changed || !State.Contributions.empty();
+  for (const GlobalObject *const G : State.Contributions) {
+    assert(isa<GlobalVariable>(GO) &&
+           "Only global variables can have contributions!");
+    assert(isa<Function>(G) && "All contributions are functions!");
+    GOHash.update(GOIMap[G].InitialDigest.Bytes);
+  }
+  // Add the GO's dependencies to its hash.
+  auto LoopPoint = std::numeric_limits<size_t>::max();
+  for (const GlobalObject *const G : State.Dependencies) {
     const llvm::Function *const Fn = dyn_cast<const llvm::Function>(G);
     // if function will not be inlined and not be discarded if it is not used,
     // skip it.
     if (Fn && Fn->hasFnAttribute(Attribute::NoInline) &&
         !Fn->isDiscardableIfUnused())
       continue;
-    Changed = updateDigestUseDependenciesAndContributions(
-                  G, GOHash, Visited, GOIMap, AccumulateGODigest) ||
-              Changed;
+    const auto GState = updateDigestUseDependenciesAndContributions(
+        G, Visited, GOIMap, GOHashCache, AccumulateGODigest);
+    Changed = Changed || std::get<0>(GState);
```

Comment:
The use of get here obscures this and doesn't help in describing what is going on here.

I think that std::tie would help. I know it would make more code, but I think that it would make the intent clearer, and this may lead to some other simplification or refactoring.

Reply:
Fixed, thanks.